### ISSUE TITLE
assignment by reference of HMM in todo

### ISSUE THREAD

**Issue by yumengch** [nansencenter/DAPPER#20]
Hi Patrick

I finally get some time to look at the DAPEPR again and I was thinking about the following issue in the TODO.

> Consider this code:
> ```python
> from dapper.mods.Lorenz63.sakov2012 import HMM1
> from dapper.mods.Lorenz63.wiljes2017 import HMM as HMM2
> ```
> Document (somewhere) that `HMM1.t` is changed by the second import, because it itself imports and changes `HMM1`. This became a bug in `test_example_2` when `pytest` was run with `--doctest-modules` because `--doctest-modules` goes through all of the modules.
> 

I don't see the mentioned problem when I run `pytest --doctest-modules`. It is possible that I use it wrong.

However, I can see why it becomes a problem due to the assignment by reference.

The goal is to make `HMM` in `dapper.mods.Lorenz63.sakov2012`  and `dapper.mods.Lorenz63.wiljes2017` two independent variable with independent reference.

A simple yet possibly inelegant method is to use `copy` in  `dapper.mods.Lorenz63.wiljes2017`:
```diff
1,40c1,43
< from dapper.mods.Lorenz63.sakov2012 import HMM, Nx
---
> from copy import copy
>
> from dapper.mods.Lorenz63.sakov2012 import HMM as _HMM
> from dapper.mods.Lorenz63.sakov2012 import Nx as _Nx
>
> Nx = _Nx
> HMM = copy(_HMM)
```
An alternative is adding a `copy` method to `HMM` (hiding the copy import in the `dapper.mods.__init__`):
```python
import copy as cp
class HiddenMarkovModel(struct_tools.NicePrint):
    ...
    def copy(self):
        return cp.copy(self)
```
This will give:
```diff
1,40c1,43
< from dapper.mods.Lorenz63.sakov2012 import HMM, Nx
---
> from dapper.mods.Lorenz63.sakov2012 import HMM as _HMM
> from dapper.mods.Lorenz63.sakov2012 import Nx as _Nx
>
> Nx = _Nx
> HMM = _HMM.copy()
```
I don't know any other way myself. Any better suggestions? I can make the changes in other mods as well if you want.

**Comment by patricknraanes** [comment]
> I finally get some time to look at the DAPEPR again and I was thinking about the following issue in the TODO.

I think it's not been very long!

> I don't see the mentioned problem when I run `pytest --doctest-modules`. It is possible that I use it wrong.

That's because the Lorenz models were excluded from pytest --doctest-modules via pyproject.toml. However, this has been updated on master since then. Now `test_example_2` sets the HMM itself instead of importing it. This avoids the bug, but is quite ugly. 

> A simple yet possibly inelegant method is to use `copy` in `dapper.mods.Lorenz63.wiljes2017`:

Yes, I've considered it, but found it too ugly. I want the mod configurations to be quick-and-dirty. Of course, that gives rise to these bugs when trying to test rigorously.

> An alternative is adding a `copy` method to `HMM` (hiding the copy import in the `dapper.mods.__init__`):

That is quite ingenious! I hadn't considered that. But:
- You'd probably need to use `deepcopy`, though. Otherwise the changes in eg `dapper.mods.Lorenz63.wiljes2017` to `HMM.t` will also be present in the copy of the `HMM`.
- IIRC, `copy`ing can run into trouble because objects cannot be `pickle`d. However, this is already a consideration for `multiprocessing`, and hasn't yielded much trouble ever since I started using `dill` (via `multiprocessing_on_dill`).
- This might have the unintended consequence that changing the forcing (e.g `dapper/mods/Lorenz96/pinheiro2019.py`) does not take effect. I'm not sure though.

In summary, this is tempting, but I'm not sure if it is feasible and practical. Please give it a think and let me know.
The alternative is again to document the fact that these settings modules are supposed to be quick-and-dirty, leading to the kind of brittleness observed with this bug.

**Comment by patricknraanes** [comment]
Closed by #22
